### PR TITLE
:seedling: (alpha): add e2e test for alpha update command

### PR DESCRIPTION
**Pull request by vitorfloriano**
This PR adds a new e2e test for the `alpha update` command that validates custom code preservation during version updates.

#### What's Added
- **New e2e test suite**: `test/e2e/alphaupdate/` with focused logging
- **Single test case**: Validates alpha update preserves custom code during scaffolding updates
- **Focused validation**: Tests core functionality with potential for future expansion

#### Test Coverage

The new test validates:
- Alpha update command execution (for now, hardcoded to v4.5.2 → v4.6.0)
- Custom API field preservation during update
- Custom controller logic preservation during update
- Project scaffolding updates correctly

#### Implementation Details
- Downloads kubebuilder v4.5.2 binary for creating the mock project
- Injects custom code into API types and controller
- Runs `alpha update` using current kubebuilder binary
- Validates all custom code remains intact after update

#### Follow-up Work
Additional validation scenarios (git branches, flags, edge cases) can be added in future iterations.

#### Test Results

- Integrated into existing e2e test suite
- Clean, readable logging output

```
=== RUN   TestE2E
  Starting kubebuilder suite test for the alpha update command
Running Suite: Kubebuilder alpha update suite - /home/vitor/go/src/github.com/vitorfloriano/kubebuilder/test/e2e/alphaupdate
============================================================================================================================
Random Seed: 1752155135

Will run 1 of 1 specs
------------------------------
kubebuilder alpha update should update project from v4.5.2 to v4.6.0 preserving custom code
/home/vitor/go/src/github.com/vitorfloriano/kubebuilder/test/e2e/alphaupdate/update_test.go:73
  STEP: setting up test context for kubebuilder binary management @ 07/10/25 10:45:35.194
  running: kubectl version -o json
  cleaning up tools
  preparing testing directory: /home/vitor/go/src/github.com/vitorfloriano/kubebuilder/test/e2e/alphaupdate/e2e-apoj
  STEP: creating isolated mock project directory in /tmp to avoid git conflicts @ 07/10/25 10:45:35.966
  STEP: downloading kubebuilder v4.5.2 binary to isolated /tmp directory @ 07/10/25 10:45:35.966
  STEP: creating mock project with kubebuilder v4.5.2 @ 07/10/25 10:45:38.773
  STEP: running kubebuilder init @ 07/10/25 10:45:38.773
  STEP: running kubebuilder create api @ 07/10/25 10:45:48.194
  STEP: running make all @ 07/10/25 10:46:00.432
  STEP: injecting custom code in API and controller @ 07/10/25 10:46:19.907
  STEP: initializing git repository and committing mock project @ 07/10/25 10:46:19.907
  STEP: initializing git repository @ 07/10/25 10:46:19.907
  STEP: adding all files to git @ 07/10/25 10:46:20.003
  STEP: committing initial project state @ 07/10/25 10:46:20.014
  STEP: running alpha update from v4.5.2 to v4.6.0 @ 07/10/25 10:46:20.032
  STEP: validating custom code preservation @ 07/10/25 10:49:11.746
  STEP: cleaning up test artifacts @ 07/10/25 10:49:11.746
  running: docker rmi -f e2e-test/controller-manager:apoj
• [217.025 seconds]
------------------------------

Ran 1 of 1 Specs in 217.027 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestE2E (217.03s)
PASS
ok      sigs.k8s.io/kubebuilder/v4/test/e2e/alphaupdate 217.038s
Deleting cluster "local-kubebuilder-e2e" ...
Deleted nodes: ["local-kubebuilder-e2e-control-plane"]
```

### CODE REVIEW

File: test/e2e/alphaupdate/update_test.go [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+	content, err = os.ReadFile(controllerFile)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(string(content)).To(ContainSubstring("Custom reconciliation logic"))
+	Expect(string(content)).To(ContainSubstring("log.Info(\"Reconciling TestOperator\")"))
+	Expect(string(content)).To(ContainSubstring("log.Info(\"TestOperator size\", \"size\", testOperator.Spec.Size)"))
+
+	projectFile := filepath.Join(projectDir, "PROJECT")
+	content, err = os.ReadFile(projectFile)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(string(content)).To(ContainSubstring("version: \"3\""))
```
Comment (camilamacedo86) [inline test/e2e/alphaupdate/update_test.go]: Could we remove it?
Does not add value, and if we will use the PROJECT file, then I think we need unmarshal and work with the struct such as: https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/alphagenerate/generate_test.go#L223-L300

Comment (camilamacedo86) [reply test/e2e/alphaupdate/update_test.go]: We test it in the alpha generation, so it is fine, we have only the above checks.

File: test/e2e/alphaupdate/update_test.go [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+			mockProjectDir  string
+			kbOldBinaryPath string
+		)
+
+		BeforeEach(func() {
+			var err error
+			By("building and installing current kubebuilder binary")
+			cmd := exec.Command("make", "install")
+			_, err = cmd.CombinedOutput()
+			Expect(err).NotTo(HaveOccurred(), "Failed to run make install")
```
Comment (camilamacedo86) [inline test/e2e/alphaupdate/update_test.go]: Could we use: 

```
output, err := cmd.CombinedOutput()
Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("make install failed: %s", string(output)))
```

TO have a better output?

File: test/e2e/alphaupdate/update_test.go [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+			Expect(err).NotTo(HaveOccurred())
+
+			By("downloading kubebuilder v4.5.2 binary to isolated /tmp directory")
+			kbOldBinaryPath, err = downloadKubebuilder()
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		AfterEach(func() {
+			// Only cleanup when running locally (not in CI/prow)
+			if os.Getenv("CI") == "" && os.Getenv("PROW_JOB_ID") == "" {
```
Comment (camilamacedo86) [inline test/e2e/alphaupdate/update_test.go]: Why not clean up in the CI?
Can we keep local and CI working ==?
In both cases, we should clean up the staff.

File: test/e2e/alphaupdate/update_test.go [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+				By("cleaning up test artifacts")
+				if mockProjectDir != "" {
+					_ = os.RemoveAll(mockProjectDir)
+				}
+				if kbOldBinaryPath != "" {
+					_ = os.RemoveAll(filepath.Dir(kbOldBinaryPath))
+				}
+
+				By("cleaning up alpha update downloaded binaries")
+				cleanupAlphaUpdateBinaries()
```
Comment (camilamacedo86) [inline test/e2e/alphaupdate/update_test.go]: I think we can remove this func and simplify it such as:

```
AfterEach(func() {
		By("cleaning up test artifacts")

		_ = os.RemoveAll(mockProjectDir)
		_ = os.RemoveAll(filepath.Dir(kbOldBinaryPath))

		// Clean up kubebuilder alpha update downloaded binaries
		binaryPatterns := []string{
			"/tmp/kubebuilder" + fromVersion + "-*", <- lets add it in var on top since it is re-used
			"/tmp/kubebuilder" + toVersion + "-*", // <- lets add it in var on top since it is re-used
		}

		for _, pattern := range binaryPatterns {
			matches, _ := filepath.Glob(pattern)
			for _, path := range matches {
				_ = os.RemoveAll(path)
			}
		}
})
```

File: test/e2e/alphaupdate/update_test.go [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+const (
+	fromVersion = "v4.5.2"
+	toVersion   = "v4.6.0"
+)
+
+var _ = Describe("kubebuilder", func() {
+	Context("alpha update", func() {
+		var (
+			mockProjectDir  string
+			kbOldBinaryPath string
```
Comment (camilamacedo86) [inline test/e2e/alphaupdate/update_test.go]: Should we not have binPath for fromVersion and toVersion?
Can we name like binFromVersionPath and binToVersionPath?

File: test/e2e/alphaupdate/update_test.go [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+		)
+
+		BeforeEach(func() {
+			var err error
+			By("building and installing current kubebuilder binary")
+			cmd := exec.Command("make", "install")
```
Comment (camilamacedo86) [inline test/e2e/alphaupdate/update_test.go]: we should not call make install? 
We have the current bin in the TestContext, we need to use it. 

https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/v4/plugin_cluster_test.go#L53-L56

File: test/e2e/alphaupdate/update_test.go (hunk 1 above)
Comment (vitorfloriano) [reply test/e2e/alphaupdate/update_test.go]: Alright, I removed that. It's redudant.

File: test/e2e/alphaupdate/update_test.go (hunk 2 above)
Comment (vitorfloriano) [reply test/e2e/alphaupdate/update_test.go]: I removed this completely since we are going to use the binary from TestContext.

File: test/e2e/alphaupdate/update_test.go (hunk 3 above)
Comment (vitorfloriano) [reply test/e2e/alphaupdate/update_test.go]: I just assumed we could skip cleanup in CI since the environment is ephemeral and everything gets discarded anyway. But I see your point about keeping behavior consistent across environments. I updated it.

File: test/e2e/alphaupdate/update_test.go (hunk 4 above)
Comment (vitorfloriano) [reply test/e2e/alphaupdate/update_test.go]: Alright. I removed that and moved those binary patterns to const at the top and simplified the AfterEach cleanup as you suggested. I also added kbc.Destroy() back in for complete cleanup.

File: test/e2e/alphaupdate/update_test.go (hunk 5 above)
Comment (vitorfloriano) [reply test/e2e/alphaupdate/update_test.go]: Alright, I changed that.

File: test/e2e/alphaupdate/update_test.go (hunk 6 above)
Comment (vitorfloriano) [reply test/e2e/alphaupdate/update_test.go]: Alright, I removed `make install` completely and switched to `utils.NewTestContext()` like in the other tests. That was the cause of the last fail in CI.

File: test/e2e/alphaupdate/e2e_suite_test.go [hunk 7]
```diff
@@ -0,0 +1,32 @@
+/*
```
Comment (camilamacedo86) [inline test/e2e/alphaupdate/e2e_suite_test.go:1]: @Vitor Floriano It looks great now!

Before merging, we just need to **squash all commits** so that there’s only **one single commit** in the PR.

Why?
- It ensures a cleaner, more understandable Git history.
- If we ever need to revert this PR, having one commit makes it much easier.
- It aligns with our golden rule:

  1 PR → 1 Problem/Purpose → 1 Commit

Let me know if you want help squashing the commits.

**Comment by camilamacedo86** [comment]
Great work 🥇 
Thank you a lot !!!

/lgtm
/approve

**Review by camilamacedo86 — APPROVED** [review]
